File: examples/python/basic.py
#!/usr/bin/env python3
"""Basic entropy collection with openentropy.

Creates an entropy pool, collects random bytes, and prints health stats.

Usage:
    pip install maturin && maturin develop --release
    python examples/python/basic.py
"""

from openentropy import EntropyPool, detect_available_sources, version

print(f"openentropy v{version()}")

# Discover what's available
sources = detect_available_sources()
print(f"\n{len(sources)} entropy sources detected:")
for s in sources:
    print(f"  - {s['name']}: {s['description']} ({s['category']})")

# Create pool with all available sources
pool = EntropyPool.auto()
print(f"\nPool created with {pool.source_count} sources")

# Collect entropy (parallel for speed)
collected = pool.collect_all(parallel=True, timeout=10.0)
print(f"Collected {collected} raw bytes")

# Get conditioned random bytes
data = pool.get_random_bytes(64)
print(f"\n64 random bytes (hex): {data.hex()}")

# Health report
report = pool.health_report()
print(f"\nHealth: {report['healthy']}/{report['total']} sources healthy")
print(f"Raw bytes in pool: {report['raw_bytes']}")
print(f"Output bytes produced: {report['output_bytes']}")

for s in report["sources"]:
    status = "✓" if s["healthy"] else "✗"
    print(f"  {status} {s['name']}: {s['bytes']} bytes, H={s['entropy']:.3f}, {s['time']:.3f}s")

File: crates/openentropy-core/examples/basic.rs
//! Basic entropy collection example.
//!
//! Auto-detects available hardware entropy sources, collects bytes,
//! and prints them as hex.
//!
//! Run: `cargo run --example basic`

use openentropy_core::EntropyPool;

fn main() {
    // Create a pool with all available sources on this machine
    let pool = EntropyPool::auto();

    println!("Sources registered: {}", pool.source_count());

    // Collect from all sources (parallel, 10s timeout)
    let bytes_collected = pool.collect_all();
    println!("Raw bytes collected: {bytes_collected}");

    // Get 64 bytes of conditioned (SHA-256) random output
    let random = pool.get_random_bytes(64);
    print!("Random bytes (hex): ");
    for b in &random {
        print!("{b:02x}");
    }
    println!();

    // Print health report
    let health = pool.health_report();
    println!(
        "\nPool health: {}/{} sources healthy, {} raw bytes total",
        health.healthy, health.total, health.raw_bytes
    );
}
